kernel/fatfs.c

```diff
@@ -921,66 +526,64 @@ static int fatfs_mount(char *source, char *tgt, uint32_t flags, void *arg)
         return -1;
     }
 
+    struct fatfs *fs = fsd->fs;
+    fs->mounted = 0;
+
     /* Associate the mount point */
     fsd->mountpoint = tgt_dir;
     tgt_dir->owner = &mod_fatfs;
+    fs->bsect = 0;
 
+    mb_read(fsd, fs->win, 0, (fs->bsect), 512);
 
-    /* Search FAT partition on the drive */
-    bsect = 0;
-    fmt = check_fs(fsd, buf, bsect);			/* Check sector 0 as an SFD format */
-    if (fmt == 1) {						/* Not an FAT boot record, it may be FDISK format */
-        /* Check a partition listed in top of the partition table */
-        if (disk_readp(fsd, buf, bsect, MBR_Table, 16)) {	/* 1st partition entry */
-            fmt = 3;
-        } else {
-            if (buf[4]) {					/* Is the partition existing? */
-                bsect = LD_DWORD(&buf[8]);	/* Partition offset in LBA */
-                fmt = check_fs(fsd, buf, bsect);	/* Check the partition */
-            }
+    if (check_fs(fsd) == -2) {
+        fs->bsect = LD_WORD(fs->win + BS_MBR + 8);
+        mb_read(fsd, fs->win, 0, (fs->bsect), 512);
```

Comment:
correct. before parsing the boot sector we can NOT know the blocksize. since we operate on sd cards which always have a sector size of 512 bytes, this is the safest assumption.